pkg/networkserver/redis/registry.go

```diff
@@ -480,305 +536,110 @@ func (r *DeviceRegistry) RangeByUplinkMatches(ctx context.Context, up *ttnpb.Upl
 	fieldKeyCurrent := FieldKey(addrKeyCurrent)
 	fieldKeyPending := FieldKey(addrKeyPending)
 
-	payloadHash := uplinkPayloadHash(up.RawPayload)
-
-	matchResultKey := ttnredis.Key(addrKey, "up", payloadHash, "result")
-	matchUIDKeyCurrentLE := ttnredis.Key(addrKeyCurrent, "up", payloadHash, "le")
-	matchUIDKeyCurrentGT := ttnredis.Key(addrKeyCurrent, "up", payloadHash, "gt")
-	matchUIDKeyPending := ttnredis.Key(addrKeyPending, "up", payloadHash)
-	matchFieldKeyCurrent := ttnredis.Key(fieldKeyCurrent, "up", payloadHash)
-	matchFieldKeyPending := ttnredis.Key(fieldKeyPending, "up", payloadHash)
-
-	var matchKeys []string
-	if pld.FHdr.FCtrl.Ack {
-		matchKeys = []string{
-			matchResultKey,
-
-			addrKeyCurrent,
-			fieldKeyCurrent,
-			matchUIDKeyCurrentLE,
-			matchUIDKeyCurrentGT,
-			matchFieldKeyCurrent,
-		}
-	} else {
-		matchKeys = []string{
-			matchResultKey,
-
-			addrKeyCurrent,
-			fieldKeyCurrent,
-			matchUIDKeyCurrentLE,
-			matchUIDKeyCurrentGT,
-			matchFieldKeyCurrent,
-
-			addrKeyPending,
-			fieldKeyPending,
-			matchUIDKeyPending,
-			matchFieldKeyPending,
+	var (
+		currentSessionScoresCmd  *redis.ZSliceCmd
+		currentSessionMappingCmd *redis.StringStringMapCmd
+		pendingSessionScoresCmd  *redis.ZSliceCmd
+		pendingSessionMappingCmd *redis.StringStringMapCmd
+	)
+	if _, err := r.Redis.TxPipelined(ctx, func(p redis.Pipeliner) error {
+		currentSessionScoresCmd = ZRangeArgsWithScores(ctx, p, addrKeyCurrent)
+		currentSessionMappingCmd = p.HGetAll(ctx, fieldKeyCurrent)
+		if !ackFlag {
+			pendingSessionScoresCmd = ZRangeArgsWithScores(ctx, p, addrKeyPending)
+			pendingSessionMappingCmd = p.HGetAll(ctx, fieldKeyPending)
 		}
+		return nil
+	}); err != nil {
+		return ttnredis.ConvertError(err)
 	}
 
-	trace.Log(ctx, "ns:redis", "run deviceMatchScript")
-	vs, err := ttnredis.RunInterfaceSliceScript(ctx, r.Redis, deviceMatchScript, matchKeys, lsb, cacheTTL.Milliseconds()).Result()
+	var (
+		currentSessionSet []sessionEntry
+		pendingSessionSet []sessionEntry
+		err               error
+	)
+	currentSessionSet, err = buildSortedSessionSet(currentSessionScoresCmd, currentSessionMappingCmd, true, lsb)
 	if err != nil {
-		if err == redis.Nil {
-			return errNoUplinkMatch.WithCause(errNoCandidates)
-		}
-		return ttnredis.ConvertError(err)
+		return err
 	}
-	if len(vs) < 1 {
-		panic("empty match script return value")
+	if !ackFlag {
+		pendingSessionSet, err = buildSortedSessionSet(pendingSessionScoresCmd, pendingSessionMappingCmd, false, 0)
+		if err != nil {
+			return err
+		}
 	}
-	matchType, ok := vs[0].(string)
-	if !ok {
-		panic(fmt.Sprintf("expected first match script return value element to be a string, got '%v'(%T)", vs[0], vs[0]))
+
+	fillContext := func(ctx context.Context, uid string) (context.Context, *ttnpb.EndDeviceIdentifiers, error) {
+		ids, err := unique.ToDeviceID(uid)
```

Comment:
We can do `WithContext` here too right?

Reply:
This actually comes with a sanity check in Enterprise.